### PR TITLE
Disable task navigation for required tasks until complete

### LINKED ISSUE #688: `required` flag isn't supported in any task types
**Package**

lib-classifier

### PR DESCRIPTION
**Pull request by eatyourgreens**
Tasks are complete if they are not required or if their annotation has a valid value.

TODO:
- [x] Add `step.isComplete` where a step is complete if all its tasks are complete.

Package:
lib-classifier

Closes #688.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
I've held off on validating drawing tasks until #1266 merges, but a drawing task is complete if all its tools are complete, which should be something like this: 
```js
self.tools.reduce( (tool, isComplete) => isComplete && tool.isComplete, true)
```

**Comment by eatyourgreens** [comment]
Tests for default behaviour are failing because the factories override default values.

File: packages/lib-classifier/test/factories/tasks/MultipleChoiceTaskFactory.js [hunk 1]
```diff
@@ -2,9 +2,9 @@ import { Factory } from 'rosie'
 
 export default new Factory()
   .attr('answers', [
-    { label: 'red', _key: Math.random() },
-    { label: 'blue', _key: Math.random() },
-    { labe: 'green', _key: Math.random() }
+    { label: 'red', _key: 1 },
+    { label: 'blue', _key: 2 },
+    { label: 'green', _key: 3 }
```
Comment (srallen) [inline packages/lib-classifier/test/factories/tasks/MultipleChoiceTaskFactory.js]: PFE uses `Math.random()` for these keys, so I'd like to use that here so the factory mocks are like what the actual tasks are (or as close as we can mock): 

https://github.com/zooniverse/Panoptes-Front-End/blob/e52b9f39a667f5e1bf4c235622e92f1a5537e2df/app/classifier/tasks/generic-editor.cjsx#L104

**Comment by srallen** [comment]
There seems to be an unrelated test for `InnerTickAxis` that seems to fail occasionally on Travis (I've never seen it fail locally). I think I need to refactor it... I'm rerunning tests now to see if everything passes.

**Comment by eatyourgreens** [comment]
Re. The multiple choice answer keys, those are integers in the model so the factory builds invalid snapshots.

**Comment by srallen** [comment]
Ok, then the type should change to `types.number`.

File: packages/lib-classifier/src/plugins/tasks/SingleChoiceTask/models/SingleChoiceTask.js [hunk 2]
```diff
@@ -12,7 +12,6 @@ const SingleChoice = types.model('SingleChoice', {
   })),
   help: types.optional(types.string, ''),
   question: types.string,
-  required: types.maybe(types.boolean), // Should this be an optional type with the default to true?
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/tasks/SingleChoiceTask/models/SingleChoiceTask.js]: You can override the base model properties, right?

File: packages/lib-classifier/src/store/Step.spec.js [hunk 3]
```diff
@@ -10,4 +15,28 @@ describe('Model > Step', function () {
...
+    it('should be complete', function () {
+      const tasks = {
+        T1: MultipleChoiceTaskFactory.build({ taskKey: 'T1', required: false }),
+        T2: SingleChoiceTaskFactory.build({ taskKey: 'T2', required: false })
+      }
+      step = Step.create({ stepKey: 'S1', taskKeys: ['T1', 'T2'], tasks })
+      step.classifications = ClassificationStore.create()
+      expect(step.isComplete).to.be.true()
+    })
+  })
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/Step.spec.js]: Good idea. There's a story for that case.

File: packages/lib-classifier/src/store/WorkflowStepStore.js [hunk 4]
```diff
@@ -23,9 +16,7 @@ const WorkflowStepStore = types
     get activeStepTasks () {
       const validStepReference = isValidReference(() => self.active)
       if (validStepReference) {
-        return self.active.taskKeys.map((taskKey) => {
-          return self.tasks.get(taskKey)
-        }).filter(Boolean)
+        return self.active.steps
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/WorkflowStepStore.js]: `self.active.tasks`. This is a typo that should have been picked up by the tests.

File: packages/lib-classifier/test/factories/tasks/MultipleChoiceTaskFactory.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/test/factories/tasks/MultipleChoiceTaskFactory.js]: Why is this model storing these React keys anyway? There are no keys stored on the single choice task answers. In the view code for the multiple choice task, the key for each answer is `${task.taskKey}_${index}`

**Comment by eatyourgreens** [comment]
>There seems to be an unrelated test for InnerTickAxis that seems to fail occasionally on Travis (I've never seen it fail locally). I think I need to refactor it... I'm rerunning tests now to see if everything passes.

I've been running into that intermittently over the last couple of weeks. Every now and again, it fails on Travis too.

File: packages/lib-classifier/test/factories/tasks/MultipleChoiceTaskFactory.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/test/factories/tasks/MultipleChoiceTaskFactory.js]: NB. underscored properties in PFE resources are ignored when saving to Panoptes.

File: packages/lib-classifier/src/store/WorkflowStepStore.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/WorkflowStepStore.js]: I had a quick look at the tests and I don't see a test for `activeStepTasks`. Should we remove it and just use `workflowSteps.active.tasks`?

File: packages/lib-classifier/src/plugins/tasks/SingleChoiceTask/models/SingleChoiceTask.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (srallen) [reply packages/lib-classifier/src/plugins/tasks/SingleChoiceTask/models/SingleChoiceTask.js]: Yeah, I mean more just the reminder comment about this. Could you open an RFC issue to keep track of it?

[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/tasks/SingleChoiceTask/models/SingleChoiceTask.js]: Sure.

File: packages/lib-classifier/src/store/WorkflowStepStore.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/WorkflowStepStore.js]: I just realised `activeStepTasks` was an array, so `step.tasks` should also be an array, not a map.

**Comment by eatyourgreens** [comment]
This works for me on I Fancy Cats. I must answer the first question before continuing. Tests are failing locally, though. 🤔

**Comment by rogerhutchings** [comment]
merge cornflakes!

also, is this ready for another pass?

**Comment by eatyourgreens** [comment]
Absolutely ready for review.

**Comment by rogerhutchings** [comment]
Would you mind resolving the merge ~~cornflake~~ conflict?

**Comment by eatyourgreens** [comment]
Excellent. Once this is merged, we can require minimum numbers of marks before being allowed to move on from a drawing task.

**Comment by eatyourgreens** [comment]
Any objections to merging this? If not, I’ll merge it tomorrow.

**Pull request merged**